src/protocol/sort/shuffle.rs

```diff
@@ -0,0 +1,280 @@
+        let (p1, p2, p3) = make_participants();
+        let random_value = rand.gen::<u64>();
+        let sequence1left = Shuffle::<Fp31, TestStep>::generate_random_sequence(
+            batchsize,
+            random_value,
+            true,
+            &p1[TestStep::Shuffle(Step1)],
+        );
+        let mut sequence1right = Shuffle::<Fp31, TestStep>::generate_random_sequence(
+            batchsize,
+            random_value,
+            false,
+            &p1[TestStep::Shuffle(Step1)],
+        );
+
+        let sequence2left = Shuffle::<Fp31, TestStep>::generate_random_sequence(
+            batchsize,
+            random_value,
+            true,
+            &p2[TestStep::Shuffle(Step1)],
+        );
+        let mut sequence2right = Shuffle::<Fp31, TestStep>::generate_random_sequence(
+            batchsize,
+            random_value,
+            false,
+            &p2[TestStep::Shuffle(Step1)],
+        );
+
+        let sequence3left = Shuffle::<Fp31, TestStep>::generate_random_sequence(
+            batchsize,
+            random_value,
+            true,
+            &p3[TestStep::Shuffle(Step1)],
+        );
+        let mut sequence3right = Shuffle::<Fp31, TestStep>::generate_random_sequence(
+            batchsize,
+            random_value,
+            false,
+            &p3[TestStep::Shuffle(Step1)],
+        );
+
+        let expected_numbers: Vec<usize> = (0..batchsize).collect();
+        assert_eq!(sequence1right, sequence2left);
+        assert_eq!(sequence2right, sequence3left);
+        assert_eq!(sequence3right, sequence1left);
+
+        sequence1right.sort_unstable();
+        sequence2right.sort_unstable();
+        sequence3right.sort_unstable();
+
+        assert_eq!(sequence1right, expected_numbers);
+        assert_eq!(sequence2right, expected_numbers);
+        assert_eq!(sequence3right, expected_numbers);
+    }
+
+    #[tokio::test]
+    async fn shuffle() {
+        let world: TestWorld<TestStep> = make_world(QueryId);
+        let context = make_contexts(&world);
+        let mut rand = StepRng::new(100, 1);
+
+        let input: HashSet<u8> = HashSet::from([4, 1, 9, 2, 3, 8, 6, 7, 5]);
+        let input_len = input.len();
+
+        let mut shares0 = Vec::with_capacity(input_len);
+        let mut shares1 = Vec::with_capacity(input_len);
+        let mut shares2 = Vec::with_capacity(input_len);
+
+        input.clone().into_iter().for_each(|iter| {
+            let share = share(Fp31::from(iter), &mut rand);
+            shares0.push(share[0]);
+            shares1.push(share[1]);
+            shares2.push(share[2]);
+        });
+
+        let mut sh0 = Shuffle::new(shares0, TestStep::Shuffle);
+        let mut sh1 = Shuffle::new(shares1, TestStep::Shuffle);
+        let mut sh2 = Shuffle::new(shares2, TestStep::Shuffle);
+
+        let shuffle0 = sh0.execute(&context[0]);
+        let shuffle1 = sh1.execute(&context[1]);
+        let shuffle2 = sh2.execute(&context[2]);
```

Comment:
These variable names could be better. What's the difference between sh0 and shuffle0? Can you please use more descriptive names?